tests/functions/test_complete.py

```diff
@@ -136,81 +167,163 @@ def test_fill_value_all_years(df1):
         }
     )
 
+
+def test_fill_value_all_years(df1, df1_output):
+    """
+    Test the complete function accurately replicates for
+    all the years from 1999 to 2004.
+    """
+
     result = df1.complete(
         columns=[
-            {"Year": range(df1.Year.min(), df1.Year.max() + 1)},
+            {"Year": lambda x: range(x.Year.min(), x.Year.max() + 1)},
             "Taxon",
         ],
         fill_value={"Abundance": 0},
     )
-    assert_frame_equal(result, output1)
+    assert_frame_equal(result, df1_output)
 
 
-def test_type_columns(df1):
-    """Raise error if columns is not a list object."""
-    with pytest.raises(TypeError):
-        df1.complete(columns="Year")
+def test_dict_series(df1, df1_output):
+    """
+    Test the complete function if a dictionary containing a Series
+    is present in `columns`.
+    """
 
+    result = df1.complete(
+        columns=[
+            {
+                "Year": lambda x: pd.Series(
+                    range(x.Year.min(), x.Year.max() + 1)
+                )
+            },
+            "Taxon",
+        ],
+        fill_value={"Abundance": 0},
+    )
+    assert_frame_equal(result, df1_output)
 
-def test_empty_columns(df1):
-    """Raise error if columns is empty"""
-    with pytest.raises(ValueError):
-        df1.complete(columns=[])
 
+def test_dict_series_duplicates(df1, df1_output):
+    """
+    Test the complete function if a dictionary containing a
+    Series (with duplicates) is present in `columns`.
+    """
 
-def test_fill_value_is_a_dict(df1):
-    """Raise error if fill_value is not a dictionary"""
-    with pytest.raises(TypeError):
-        df1.complete(columns=["Year", "Taxon"], fill_value=0)
+    result = df1.complete(
+        columns=[
+            {
+                "Year": pd.Series(
+                    [1999, 2000, 2000, 2001, 2002, 2002, 2002, 2003, 2004]
+                )
+            },
+            "Taxon",
+        ],
+        fill_value={"Abundance": 0},
+    )
+    assert_frame_equal(result, df1_output)
 
 
-frame = pd.DataFrame(
+# from https://tidyr.tidyverse.org/reference/complete.html
+df = pd.DataFrame(
```

Comment:
Do you intend for this dataframe to be a fixture instead?

Reply:
I used it in a parametrization. Reordered the data though; hopefully, it is clearer.